packages/world/test.ts

```diff
@@ -45,45 +46,81 @@ describe("WorldReader", () => {
         });
-        const missing = new Set();
-        if ("Palette" in section) {
-            for (let x = 0; x < 16; x++) {
-                for (let y = 0; y < 16; y++) {
-                    for (let z = 0; z < 16; z++) {
+    });
+    describe("#getBlockStateAt", () => {
+        test("should be able to seek legacy version chunk", async () => {
+            const expected: any = {};
+            readFileSync(`${root}/saves/1.12.2/mock.txt`).toString().split("\n").map((l) => l.split(" ")).forEach(([id, x, y, z, name]) => {
+                if (typeof x === "undefined") { return; }
+                let i = (Number.parseInt(y) << 8) | (Number.parseInt(z) << 4) | (Number.parseInt(x));
+                expected[i] = Number.parseInt(id);
+            });
+            const reader = await WorldReader.create(`${root}/saves/1.12.2`);
+            const region = await reader.getRegionData(0, 0);
+            const chunkY = 0;
+            const section = RegionReader.getSection(region, chunkY);
+            for (let x = 0; x < 16; x += 2) {
+                for (let y = 0; y < 16; y += 2) {
+                    for (let z = 0; z < 16; z += 2) {
                         let i = y << 8 | z << 4 | x;
-                        const result = RegionReader.seekBlockState(section, i);
-                        if (!result) {
-                            missing.add({ pos: { x, y: y + chunkY * 16, z }, id: RegionReader.seekBlockId(section, i) });
-                        } else if (result && result.Name !== "minecraft:air" && result.Name !== "minecraft:bedrock") {
-                            // console.log(`(${x}, ${y}, ${z}): ${result.Name}`);
+                        const id = RegionReader.seekBlockStateId(section, i);
+                        const expectedId = expected[i];
+                        if (typeof expectedId === "undefined") {
+                            continue;
                         }
+                        expect(id).toEqual(expectedId);
                     }
                 }
             }
-
-            if (missing.size !== 0) {
-                // console.log(`Level ${chunkY}`);
-                // console.log(missing);
-                // console.log(section.Palette);
-                // console.log(section.Palette.length);
+        });
+        test("should be able to seek new version chunk", async () => {
+            const reader = await WorldReader.create(`${root}/saves/1.14.4`);
+            const region = await reader.getRegionData(0, 0);
+            const chunkY = 0;
+            const section = RegionReader.getSection(region, chunkY) as NewRegionSectionDataFrame;
+            const expected: any = {};
+            readFileSync(`${root}/saves/1.14.4/mock.txt`).toString().split("\n").map((l) => l.split(" ")).forEach(([id, x, y, z, name]) => {
+                if (typeof x === "undefined") { return; }
+                let i = (Number.parseInt(y) << 8) | (Number.parseInt(z) << 4) | (Number.parseInt(x));
```

Comment:
Bitwise operations should be more performant than plain division/multiplication. But since this code is only executed once, it should not make a difference.

Reply:
I create function for these bit operation to improve the readability

Reply:
@lukechu10 as far as I know this is not the case due to compiler optimization. I'd be surprised if you'd even notice the tiniest of improvement in a high-level (interpreted?) language like JS.

Reply:
> @lukechu10 as far as I know this is not the case due to compiler optimization. I'd be surprised if you'd even notice the tiniest of improvement in a high-level (interpreted?) language like JS.

Honestly, I don't know. I used to program a lot in C++ and this was the type of things we would do.